lib/url.ts

```diff
@@ -0,0 +1,512 @@
+      this._query = query;
+    } else {
+      this.set(query, URLTokenizerState.QUERY);
+    }
+  }
+
+  /**
+   * Set the parts of this URL by parsing the provided text using the provided startState.
+   */
+  private set(text: string, startState: URLTokenizerState): void {
+    const tokenizer = new URLTokenizer(text, startState);
+
+    while (tokenizer.next()) {
+      const token: URLToken | undefined = tokenizer.current();
+      if (token) {
+        switch (token.type) {
+          case URLTokenType.SCHEME:
+            this._scheme = token.text || undefined;
+            break;
+
+          case URLTokenType.HOST:
+            this._host = token.text || undefined;
+            break;
+
+          case URLTokenType.PORT:
+            this._port = token.text || undefined;
+            break;
+
+          case URLTokenType.PATH:
+            const tokenPath: string | undefined = token.text || undefined;
+            if (!this._path || this._path === "/" || tokenPath !== "/") {
+              this._path = tokenPath;
+            }
+            break;
+
+          case URLTokenType.QUERY:
+            let queryString: string | undefined = token.text;
+            if (queryString) {
+              if (queryString.startsWith("?")) {
+                queryString = queryString.substring(1);
+              }
+
+              for (const queryParameterString of queryString.split("&")) {
+                const queryParameterParts: string[] = queryParameterString.split("=");
+                if (queryParameterParts.length === 2) {
+                  this.setQueryParameter(queryParameterParts[0], queryParameterParts[1]);
+                } else {
+                  throw new Error("Malformed query parameter: " + queryParameterString);
+                }
+              }
+            }
+            break;
+
+          default:
+            throw new Error(`Unrecognized URLTokenType: ${token.type} (${URLTokenType[token.type]})`);
+        }
+      }
+    }
+  }
+
+  public toString(): string {
+    let result = "";
+
+    if (this._scheme) {
+      result += `${this._scheme}://`;
+    }
+
+    if (this._host) {
+      result += this._host;
+    }
+
+    if (this._port) {
+      result += `:${this._port}`;
+    }
+
+    if (this._path) {
+      if (!this._path.startsWith("/")) {
+        result += "/";
+      }
+      result += this._path;
+    }
+
+    if (this._query) {
+      let queryString = "";
+      for (const queryParameterName in this._query) {
+        if (queryString) {
+          queryString += "&";
+        }
+        queryString += `${queryParameterName}=${this._query[queryParameterName]}`;
+      }
+      if (queryString) {
+        result += `?${queryString}`;
+      }
+    }
+
+    return result;
+  }
+
+  public static parse(text: string): URLBuilder {
+    const result = new URLBuilder();
+    result.set(text, URLTokenizerState.SCHEME_OR_HOST);
+    return result;
+  }
+}
+
+export enum URLTokenizerState {
```

Comment:
Really small thing but may want to consider using [const enums](https://basarat.gitbooks.io/typescript/content/docs/enums.html#const-enums) for space

Reply:
In fact don't even worry about that in this PR.. let's think about space stuff down the line

Reply:
Too late. I've already done it. I'm just running it through my tests to make sure everything still works.